statistics/table.go

```diff
@@ -611,3 +612,50 @@ func getPseudoRowCountByUnsignedIntRanges(intRanges []*ranger.Range, tableRowCou
 	}
 	return rowCount
 }
+
+func getPseudoColSize(tp *types.FieldType) float64 {
+	if tp == nil {
+		return pseudoColSize
+	}
+	switch tp.Tp {
+	case mysql.TypeFloat:
+		return 4
+	case mysql.TypeTiny, mysql.TypeShort, mysql.TypeInt24, mysql.TypeLong, mysql.TypeLonglong,
+		mysql.TypeDouble, mysql.TypeYear:
+		return 8
+	case mysql.TypeDuration, mysql.TypeDate, mysql.TypeDatetime, mysql.TypeTimestamp:
+		return 16
+	case mysql.TypeNewDecimal:
+		return types.MyDecimalStructSize
+	default:
+		return pseudoColSize
+	}
+}
+
+// GetAvgRowSize computes average row size for given columns.
+func (coll *HistColl) GetAvgRowSize(cols []*expression.Column, isEncodedKey bool) (size float64) {
+	if coll.Pseudo || len(coll.Columns) == 0 || coll.Count == 0 {
+		for _, col := range cols {
+			size += getPseudoColSize(col.RetType)
+		}
+	} else {
+		for _, col := range cols {
+			colHist, ok := coll.Columns[col.UniqueID]
+			// Normally this would not happen, it is for compatibility with old version stats which
+			// does not include TotColSize.
+			if !ok || (colHist.TotColSize == 0 && (colHist.NullCount != coll.Count)) {
+				size += getPseudoColSize(col.RetType)
+				continue
+			}
+			// We differentiate if the column is encoded as key or value, because the resulted size
+			// is different.
+			if isEncodedKey {
+				size += colHist.AvgColSize(coll.Count)
+			} else {
+				size += float64(colHist.TotColSize) / float64(coll.Count)
```

Comment:
Note that the `Count` is maintained for DML, while the `TotColSize` is not for fixed length columns.

Reply:
Could you elaborate more on the comments? From my observation, `TotColSize` is maintained for fixed length columns actually?

Reply:
In https://github.com/pingcap/tidb/blob/master/table/tables/tables.go#L342, the `GetBytes` for fixed length columns will always be 0, so they will not be maintained.